src/persist/src/indexed/trace.rs

```diff
@@ -108,19 +111,39 @@ impl<K: Data, V: Data> BlobTrace<K, V> {
         BlobTraceMeta {
             id: self.id,
             since: self.since.clone(),
+            seal: self.seal.clone(),
             batches: self.batches.clone(),
             next_blob_id: self.next_blob_id,
         }
     }
 
     /// An upper bound on the times of contained updates.
+    ///
+    /// Invariant:
+    /// - self.ts_upper() <= self.seal()
     pub fn ts_upper(&self) -> Antichain<u64> {
```

Comment:
This is a great question!

Added this response as a comment

```
122     /// While `self.seal` tracks the frontier of times that have been logically been
123     /// closed and are eligible to be moved into the trace, `self.ts_upper()` tracks
124     /// the frontier of times that have actually been physically moved into the trace.
125     /// `self.seal()` is required to manage invariants between commands (e.g. a seal request
126     /// has to be at a time in advance of prior seal requests) whereas `self.ts_upper()`
127     /// is required to manage physical reads and writes to the trace (e.g. to determine
128     /// which times may be added that are not already present.
```

Reply:
Makes sense! Great comment :clap:. And yeah, any time I ask a question like this in code review, it's probably a good prompt to add a comment in case someone else later has the same question. (This is something I try to do personally)